### PR TITLE
Fixes #6568 Add failure analyser for ambiguous configuration of Hikar…

### LINKED ISSUE #6568: dataSourceClassName can not be configured for HikariCP
In Spring Boot 1.4 configuration of `data-source-class-name` property for HikariCP results in runtime exception. Example configuration:
```yml
spring:
  datasource:
    url: jdbc:postgresql://localhost/database-name
    hikari:
      data-source-class-name: org.postgresql.ds.PGSimpleDataSource
```
Then getting exception like this on application startup:
```
Caused by: java.lang.IllegalArgumentException: cannot use driverClassName and dataSourceClassName together.
	at com.zaxxer.hikari.HikariConfig.validate(HikariConfig.java:770) ~[HikariCP-2.4.7.jar:na]
```
The reason of such behavior is in `org.springframework.boot.autoconfigure.jdbc.DataSourceConfiguration` which unconditionally configures `driverClassName` for any dataSource in `createDataSource` method.

### PR DESCRIPTION
**Pull request by brettwooldridge**
…iCP.  Driver or DataSource?  Both cannot be configured at the same time.

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@brettwooldridge Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-boot&pullRequestId=6716)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-boot&pullRequestId=6716) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/faq) for frequently asked questions.

File: spring-boot/src/main/resources/META-INF/spring.factories [hunk 1]
```diff
@@ -38,7 +38,8 @@ org.springframework.boot.diagnostics.analyzer.BeanNotOfRequiredTypeFailureAnalyz
 org.springframework.boot.diagnostics.analyzer.BindFailureAnalyzer,\
 org.springframework.boot.diagnostics.analyzer.NoUniqueBeanDefinitionFailureAnalyzer,\
 org.springframework.boot.diagnostics.analyzer.PortInUseFailureAnalyzer,\
-org.springframework.boot.diagnostics.analyzer.ValidationExceptionFailureAnalyzer
+org.springframework.boot.diagnostics.analyzer.ValidationExceptionFailureAnalyzer,\
+org.springframework.boot.diagnostics.analyzer.HikariAmbiguousDriverFailureAnalyzer
```
Comment (snicoll) [inline spring-boot/src/main/resources/META-INF/spring.factories]: We order those lexicographically.

**Comment by snicoll** [comment]
Can we have a test for it please?

**Comment by brettwooldridge** [comment]
I hate to be a dick, but I don't sign contributor license agreements.  The code is, as acknowledged in the header, made available under the Apache License, version 2.0.  In case you haven't read it, it specifically covers:

> 2. Grant of Copyright License. Subject to the terms and conditions of this License, each Contributor hereby grants to You a perpetual, worldwide, non-exclusive, no-charge, royalty-free, irrevocable copyright license to reproduce, prepare Derivative Works of, publicly display, publicly perform, sublicense, and distribute the Work and such Derivative Works in Source or Object form.

> 3. Grant of Patent License. Subject to the terms and conditions of this License, each Contributor hereby grants to You a perpetual, worldwide, non-exclusive, no-charge, royalty-free, irrevocable (except as stated in this section) patent license to make, have made, use, offer to sell, sell, import, and otherwise transfer the Work, where such license applies only to those patent claims licensable by such Contributor that are necessarily infringed by their Contribution(s) alone or by combination of their Contribution(s) with the Work to which such Contribution(s) was submitted.

> 5. Submission of Contributions. Unless You explicitly state otherwise, any Contribution intentionally submitted for inclusion in the Work by You to the Licensor shall be under the terms and conditions of this License, without any additional terms or conditions.

That pretty well covers it.  You are free to take the submission, or not.

**Comment by brettwooldridge** [comment]
@snicoll I'll see if I can come up with a test.  Not familiar with your test semantics (mocks, no mocks?) so it'll take a few hours.  Unfortunately, it is 12:30am here in Tokyo, so tomorrow...

**Comment by snicoll** [comment]
> I don't sign contributor license agreements

Sorry Brett but we'll have to pass on this one then. Thank you for offering the PR!

**Comment by brettwooldridge** [comment]
No problem.  Feel free to riff off of it if you think it benefits your users.

**Pull request closed without merging**